File: apps/workflow-engine/scripts/mock-llm-server.py
"""
Mock LLM Proxy Server for testing

This script creates a local server that mimics the LLM proxy endpoint.
By default it returns mock responses. Set GOOGLE_AI_API_KEY to use real Gemini.
If API fails, returns a dummy response indicating the failure.

Usage:
  python scripts/mock-llm-server.py                    # Mock responses
  GOOGLE_AI_API_KEY=your-key python scripts/mock-llm-server.py  # Real Gemini

Then test with:
  curl -X POST http://localhost:8000/run_llm \
    -H 'Content-Type: application/json' \
    -d '{"system_prompt": "You are helpful", "user_prompt": "Hello", "temperature": 0.7, "max_output_tokens": 100}'
"""

import os
import json
from http.server import HTTPServer, BaseHTTPRequestHandler
import urllib.request
import urllib.error

PORT = 8000
GEMINI_MODEL = os.environ.get('GEMINI_MODEL', 'gemini-2.0-flash-lite')
GOOGLE_AI_API_KEY = os.environ.get('GOOGLE_AI_API_KEY')
USE_MOCK = not GOOGLE_AI_API_KEY


def mock_response(req: dict) -> dict:
    user_prompt = req.get('user_prompt', '')
    system_prompt = req.get('system_prompt', '')
    temperature = req.get('temperature', 0.7)
    max_tokens = req.get('max_output_tokens', 100)

    mock_text = f"""Mock response to: "{user_prompt[:50]}..."

This is a mock response from the test server. The LLM proxy is working correctly!

System context was: {(system_prompt[:30] + '...') if system_prompt else '(none)'}
Temperature: {temperature}, Max tokens: {max_tokens}"""

    return {
        'response': mock_text,
        'usage': {
            'prompt_tokens': len(user_prompt) // 4,
            'completion_tokens': len(mock_text) // 4,
            'total_tokens': (len(user_prompt) + len(mock_text)) // 4,
        }
    }


def api_failed_response(error_msg: str) -> dict:
    """Return a dummy response when API fails"""
    return {
        'response': f"[API FAILED - DUMMY RESPONSE] The API call failed with error: {error_msg}. This is a mock response for testing the flow.",
        'usage': {
            'prompt_tokens': 0,
            'completion_tokens': 0,
            'total_tokens': 0,
        }
    }


def call_gemini(req: dict) -> dict:
    if USE_MOCK:
        return mock_response(req)

    url = f"https://generativelanguage.googleapis.com/v1beta/models/{GEMINI_MODEL}:generateContent?key={GOOGLE_AI_API_KEY}"

    request_body = {
        'contents': [
            {
                'role': 'user',
                'parts': [{'text': req.get('user_prompt', '')}]
            }
        ],
        'generationConfig': {
            'temperature': req.get('temperature', 0.7),
            'maxOutputTokens': req.get('max_output_tokens', 100)
        }
    }

    if req.get('system_prompt'):
        request_body['systemInstruction'] = {
            'parts': [{'text': req['system_prompt']}]
        }

    try:
        http_req = urllib.request.Request(
            url,
            data=json.dumps(request_body).encode('utf-8'),
            headers={'Content-Type': 'application/json'},
            method='POST'
        )

        with urllib.request.urlopen(http_req, timeout=30) as response:
            data = json.loads(response.read().decode('utf-8'))

        if not data.get('candidates') or len(data['candidates']) == 0:
            return api_failed_response('No response from Gemini API')

        text = ''.join(
            p.get('text', '') for p in data['candidates'][0]['content']['parts']
        )

        return {
            'response': text,
            'usage': {
                'prompt_tokens': data.get('usageMetadata', {}).get('promptTokenCount', 0),
                'completion_tokens': data.get('usageMetadata', {}).get('candidatesTokenCount', 0),
                'total_tokens': data.get('usageMetadata', {}).get('totalTokenCount', 0),
            }
        }

    except urllib.error.HTTPError as e:
        error_msg = f"Gemini API error ({e.code}): {e.read().decode('utf-8')}"
        print(f"API Error: {error_msg}")
        return api_failed_response(error_msg)
    except urllib.error.URLError as e:
        error_msg = f"Network error: {e.reason}"
        print(f"API Error: {error_msg}")
        return api_failed_response(error_msg)
    except Exception as e:
        error_msg = str(e)
        print(f"API Error: {error_msg}")
        return api_failed_response(error_msg)


class LLMHandler(BaseHTTPRequestHandler):
    def _set_cors_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Accept')

    def do_OPTIONS(self):
        self.send_response(204)
        self._set_cors_headers()
        self.end_headers()

    def do_POST(self):
        if self.path == '/run_llm':
            try:
                content_length = int(self.headers.get('Content-Length', 0))
                body = self.rfile.read(content_length).decode('utf-8')
                llm_request = json.loads(body)

                print(f"Received request: system_prompt={llm_request.get('system_prompt', '')[:50]}..., "
                      f"user_prompt={llm_request.get('user_prompt', '')[:50]}..., "
                      f"temperature={llm_request.get('temperature')}, "
                      f"max_output_tokens={llm_request.get('max_output_tokens')}")

                result = call_gemini(llm_request)

                print(f"Response: {result['response'][:100]}...")

                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self._set_cors_headers()
                self.end_headers()
                self.wfile.write(json.dumps(result).encode('utf-8'))

            except Exception as e:
                print(f"Error: {e}")
                self.send_response(200)  # Still return 200 with dummy response
                self.send_header('Content-Type', 'application/json')
                self._set_cors_headers()
                self.end_headers()
                result = api_failed_response(str(e))
                self.wfile.write(json.dumps(result).encode('utf-8'))
        else:
            self.send_response(404)
            self.send_header('Content-Type', 'application/json')
            self._set_cors_headers()
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'Not found'}).encode('utf-8'))

    def log_message(self, format, *args):
        # Suppress default logging
        pass


def main():
    server = HTTPServer(('', PORT), LLMHandler)
    print(f"Mock LLM Proxy Server running at http://localhost:{PORT}")
    print(f"Mode: {'MOCK (no API key)' if USE_MOCK else 'REAL (using Gemini API)'}")
    print(f"Endpoint: POST http://localhost:{PORT}/run_llm")
    print()
    print("Test with:")
    print(f"  curl -X POST http://localhost:{PORT}/run_llm \\")
    print("    -H 'Content-Type: application/json' \\")
    print('    -d \'{"system_prompt": "You are helpful", "user_prompt": "Say hello", "temperature": 0.7, "max_output_tokens": 100}\'')
    print()
    print("Note: If API fails, server returns a dummy response for flow testing.")

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nShutting down...")
        server.shutdown()


if __name__ == '__main__':
    main()

File: apps/workflow-engine/scripts/mock-llm-server.ts
/**
 * Mock LLM Proxy Server for testing
 *
 * This script creates a local server that mimics the LLM proxy endpoint.
 * By default it returns mock responses. Set GOOGLE_AI_API_KEY to use real Gemini.
 *
 * Usage:
 *   npx tsx scripts/mock-llm-server.ts                    # Mock responses
 *   GOOGLE_AI_API_KEY=your-key npx tsx scripts/mock-llm-server.ts  # Real Gemini
 *
 * Then test with:
 *   curl -X POST http://localhost:8000/run_llm \
 *     -H 'Content-Type: application/json' \
 *     -d '{"system_prompt": "You are helpful", "user_prompt": "Hello", "temperature": 0.7, "max_output_tokens": 100}'
 */

import { createServer, IncomingMessage, ServerResponse } from 'http';

const PORT = 8000;
const GEMINI_MODEL = process.env.GEMINI_MODEL || 'gemini-2.0-flash-lite';
const USE_MOCK = !process.env.GOOGLE_AI_API_KEY;

interface LLMRequest {
  system_prompt: string;
  user_prompt: string;
  temperature: number;
  max_output_tokens: number;
}

interface LLMResponse {
  response: string;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

function mockResponse(req: LLMRequest): LLMResponse {
  const mockText = `Mock response to: "${req.user_prompt.substring(0, 50)}..."

This is a mock response from the test server. The LLM proxy is working correctly!

System context was: ${req.system_prompt ? req.system_prompt.substring(0, 30) + '...' : '(none)'}
Temperature: ${req.temperature}, Max tokens: ${req.max_output_tokens}`;

  return {
    response: mockText,
    usage: {
      prompt_tokens: Math.floor(req.user_prompt.length / 4),
      completion_tokens: Math.floor(mockText.length / 4),
      total_tokens: Math.floor((req.user_prompt.length + mockText.length) / 4),
    },
  };
}

async function callGemini(req: LLMRequest): Promise<LLMResponse> {
  if (USE_MOCK) {
    return mockResponse(req);
  }

  const apiKey = process.env.GOOGLE_AI_API_KEY!;

  const url = `https://generativelanguage.googleapis.com/v1beta/models/${GEMINI_MODEL}:generateContent?key=${apiKey}`;

  const requestBody: Record<string, unknown> = {
    contents: [
      {
        role: 'user',
        parts: [{ text: req.user_prompt }],
      },
    ],
    generationConfig: {
      temperature: req.temperature,
      maxOutputTokens: req.max_output_tokens,
    },
  };

  if (req.system_prompt) {
    requestBody.systemInstruction = {
      parts: [{ text: req.system_prompt }],
    };
  }

  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(requestBody),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Gemini API error (${response.status}): ${error}`);
  }

  const data = await response.json();

  if (!data.candidates || data.candidates.length === 0) {
    throw new Error('No response from Gemini API');
  }

  const text = data.candidates[0].content.parts
    .map((p: { text: string }) => p.text)
    .join('');

  return {
    response: text,
    usage: {
      prompt_tokens: data.usageMetadata?.promptTokenCount || 0,
      completion_tokens: data.usageMetadata?.candidatesTokenCount || 0,
      total_tokens: data.usageMetadata?.totalTokenCount || 0,
    },
  };
}

function parseBody(req: IncomingMessage): Promise<string> {
  return new Promise((resolve, reject) => {
    let body = '';
    req.on('data', (chunk) => (body += chunk));
    req.on('end', () => resolve(body));
    req.on('error', reject);
  });
}

const server = createServer(async (req: IncomingMessage, res: ServerResponse) => {
  // CORS headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Accept');

  if (req.method === 'OPTIONS') {
    res.writeHead(204);
    res.end();
    return;
  }

  if (req.method === 'POST' && req.url === '/run_llm') {
    try {
      const body = await parseBody(req);
      const llmRequest: LLMRequest = JSON.parse(body);

      console.log('Received request:', {
        system_prompt: llmRequest.system_prompt?.substring(0, 50) + '...',
        user_prompt: llmRequest.user_prompt?.substring(0, 50) + '...',
        temperature: llmRequest.temperature,
        max_output_tokens: llmRequest.max_output_tokens,
      });

      const result = await callGemini(llmRequest);

      console.log('Response:', result.response.substring(0, 100) + '...');

      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(result));
    } catch (error) {
      console.error('Error:', error);
      res.writeHead(500, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: String(error) }));
    }
    return;
  }

  res.writeHead(404, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify({ error: 'Not found' }));
});

server.listen(PORT, () => {
  console.log(`Mock LLM Proxy Server running at http://localhost:${PORT}`);
  console.log(`Mode: ${USE_MOCK ? 'MOCK (no API key)' : 'REAL (using Gemini API)'}`);
  console.log(`Endpoint: POST http://localhost:${PORT}/run_llm`);
  console.log('');
  console.log('Test with:');
  console.log(`  curl -X POST http://localhost:${PORT}/run_llm \\`);
  console.log("    -H 'Content-Type: application/json' \\");
  console.log('    -d \'{"system_prompt": "You are helpful", "user_prompt": "Say hello", "temperature": 0.7, "max_output_tokens": 100}\'');
});
